### ISSUE TITLE
SO list offsets are wrong/counterintuitive

### ISSUE THREAD

**Issue by bwvdnbro** [ICRAR/VELOCIraptor-STF#108]
**Describe the bug**
The offsets stored in the SO list output seem not consistent with the SO sizes in the same file. My (possibly wrong) expectation is that the particle IDs belonging to SO `i` are stored in `pIDs[offset[i] : offset[i]+size[i]]`, but that is not the case.

**To Reproduce**
The problem can be best illustrated through the following Python snippet that can be applied to any HDF5 SO list output:
```
import h5py
SOfile = h5py.File("<RANDOM SO_LIST FILE.hdf5>", "r")
ofs = SOfile["Offset"][:]
siz = SOfile["SO_size"][:]
pIDs = SOfile["Particle_IDs"][:]
if not ofs[-1]+siz[-1] == pIDs.shape[0]:
  print("Wrong final size ({0}=/={1})!".format(ofs[-1]+siz[-1], pIDs.shape[0]))
for i in range(1,ofs.shape[0]):
  if not ofs[i-1]+siz[i-1] == ofs[i]:
    print("Wrong offset ({0} {1}, {2} {3})!".format(ofs[i-1], siz[i-1], ofs[i], siz[i]))
    exit(1)
```
This will produce two error messages. The first one because the final offset is not separated from the end of the particle ID list by the size of the final SO, and the second because the next offset does not match the previous offset plus the previous size.

**Expected behavior**
`ofs[i] == ofs[i-1]+siz[i-1]`
Put differently, I would expect `ofs` to be equivalent to
```
ofs = numpy.cumsum(siz)
ofs[1:] = ofs[:-1]
ofs[0] = 0
```

**Log files**
Not applicable.

**Environment (please complete the following information):**
Irrelevant for this problem.

**Additional context**
I think the problem is situated in `io.cxx:1450`, where the value of `SOpids[0]` seems to be (incorrectly) omitted from the loop.

**Comment by rtobar** [comment]
Thanks @bwvdnbro for recording the issue here.

tl;dr: you are almost right: yes, there is an element being skipped when building `offset` from `SOpids`, but the missing index is `1`, not `0`. We can quickly patch this now, or you can wait a bit more for a more complete solution.

The longer story now: like I mentioned in Slack, I had actually been looking at this code these days in the context of trying to reduce the maximum memory usage (see #71). The code has some extra complexity because the `SOpids` and `SOtypes` arrays-of-vectors are originally allocated in `GetSOMasses` with `N + 1` elements:

https://github.com/ICRAR/VELOCIraptor-STF/blob/b3c337179f1d24884addfd570e4a631e697d6b7f/src/substructureproperties.cxx#L3306-L3311

Then they are 1-indexed:

https://github.com/ICRAR/VELOCIraptor-STF/blob/b3c337179f1d24884addfd570e4a631e697d6b7f/src/substructureproperties.cxx#L3372

https://github.com/ICRAR/VELOCIraptor-STF/blob/b3c337179f1d24884addfd570e4a631e697d6b7f/src/substructureproperties.cxx#L3751-L3758

The same repeats in `GetInclusiveMasses`.

Skipping element `0` is then fine, because it does't contain anything. What *is* being skipped though is the size of element `1` in the loop populating `offset`:

https://github.com/ICRAR/VELOCIraptor-STF/blob/b3c337179f1d24884addfd570e4a631e697d6b7f/src/io.cxx#L1461

In that line the indexing should happen with `SOpids[i - 1]`.

As part of the changes that I've been working on for #71 I've also changed the indexing strategy for `SOpids` and `SOtypes` to be 0-based, which in turn makes their manipulation much clearer in other places. My changes have also fixed this problem actually, although I hadn't noticed them before you raised this issue. If you are in a hurry we can quickly try a patch now, but if you're happy to wait a bit more for the more complete solution then a fix for this problem will be included when I merge my changes.

**Comment by bwvdnbro** [comment]
Sorry, I meant element 1 and not 0 (my mind is just not used to thinking about 1 as the first element).

There is absolutely no hurry, I am currently using the `numpy.cumsum` workaround in my script.

**Comment by rtobar** [comment]
@bwvdnbro please have a look at the changes in the `issue-71` branch, that should now yield a `Offset` dataset. Please let me know if you see now the correct values (and that nothing else breaks), and after confirmation I'll merge this to the `master`.

**Comment by bwvdnbro** [comment]
@rtobar, I can now confirm that the changes in the `issue-71` fix the offsets and do not affect any of the other quantities we look at.

**Comment by rtobar** [comment]
Branch merged now, thanks for testing.
